### ISSUE TITLE
Livepatch does not handle static keys

### ISSUE THREAD

**Issue by euspectre** [dynup/kpatch#931]
TL;DR:
If the patched function has a condition with [static keys](https://www.kernel.org/doc/Documentation/static-keys.txt), the condition will always remain in its initial state there. It will not be switched by `static_key_slow_inc()/static_key_slow_dec()`. A workaround is to check the condition with `static_key_enabled()` in the patched code.

kpatch-build should at least detect such conditions and fail the build of the patch or build the patch so that the sure static keys are handled automatically.

---

Here is the long story.

While we were testing the live patch with [the fix for CVE-2018-9568](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/net/core/sock.c?id=9d538fa60bad4f7b23193c89e843797a1cf71ef3) for Virtuozzo 7, some of the tests caught kernel crashes due to incorrect reference counts of cgroups.

It took a while to debug but finally the root cause was found.
I reproduced the problem in CentOS 7.6 with the kernel 3.10.0-957.1.3.el7.

The source patch looked OK, it just added `newsk->sk_prot_creator = sk->sk_prot;` in `sk_clone_lock()`. It was far from obvious that `sk_update_clone()` was inlined into that function and made it into the patched `sk_clone_lock()`. It was even less obvious that sk_update_clone() used a static key condition:

net/core/sock.c:
```
static void sk_update_clone(const struct sock *sk, struct sock *newsk)
{
	if (mem_cgroup_sockets_enabled && sk->sk_cgrp) // here
		sock_update_memcg(newsk);
```
`mem_cgroup_sockets_enabled` is actually not a variable:
include/net/sock.h:
```
#define mem_cgroup_sockets_enabled static_key_false(&memcg_socket_limit_enabled)
```
The default state of the condition `if (mem_cgroup_sockets_enabled)` translates to a nop in the binary code:
```
   0xffffffffa82201fb:  0f 1f 44 00 00  nopl   0x0(%rax,%rax,1)  // here
   0xffffffffa8220200:  48 8b 43 28     mov    0x28(%rbx),%rax
<...>
   // the rest of the conditional, this part is not used by default
   0xffffffffa8220248:  49 83 bc 24 88 02 00 00 00      cmpq   $0x0,0x288(%r12)
   0xffffffffa8220251:  74 ad   je     0xffffffffa8220200
   0xffffffffa8220253:  48 89 df        mov    %rbx,%rdi
   0xffffffffa8220256:  e8 c5 06 c1 ff  callq  0xffffffffa7e30920  // sock_update_memcg
   0xffffffffa822025b:  eb a3   jmp    0xffffffffa8220200
```
When the static key is enabled by setting a limit via `memory.kmem.tcp.limit_in_bytes` in the memory cgroup, nop becomes a jump to the rest of the conditional:
```
   0xffffffffa82201fb:  e9 48 00 00 00  jmpq   0xffffffffa8220248  // here
   0xffffffffa8220200:  48 8b 43 28     mov    0x28(%rbx),%rax
<...>
   // the rest of the conditional is executed in this case
   0xffffffffa8220248:  49 83 bc 24 88 02 00 00 00      cmpq   $0x0,0x288(%r12)
   0xffffffffa8220251:  74 ad   je     0xffffffffa8220200
   0xffffffffa8220253:  48 89 df        mov    %rbx,%rdi
   0xffffffffa8220256:  e8 c5 06 c1 ff  callq  0xffffffffa7e30920  // sock_update_memcg
   0xffffffffa822025b:  eb a3   jmp    0xffffffffa8220200
```
This is how everything works for the original (not patched) binary code. So far so good.

The condition in `sk_clone_lock()` from the livepatch, however, remains in its default, disabled, state forever (checked that in the binary code of the loaded patch).

As a result, `sock_update_memcg()` will not be called for the given sockets, so `mem_cgroup_get(memcg)` will not be called either (see mm/memcontrol.c). 
```
void sock_update_memcg(struct sock *sk)
{
	if (mem_cgroup_sockets_enabled) {
<...>
		if (!mem_cgroup_is_root(memcg) && memcg_proto_active(cg_proto)) {
			mem_cgroup_get(memcg);
			sk->sk_cgrp = cg_proto;
		}
<...>
```
However, `sock_release_memcg()` is not patched, so `mem_cgroup_put(memcg)` will be called in time. 
```
void sock_release_memcg(struct sock *sk)
{
	if (mem_cgroup_sockets_enabled && sk->sk_cgrp) {
		struct mem_cgroup *memcg;
		WARN_ON(!sk->sk_cgrp->memcg);
		memcg = sk->sk_cgrp->memcg;
		mem_cgroup_put(memcg); // here
	}
}
```
Put() without get() => broken reference count and potential memory corruption.

---

To demonstrate the problem in CentOS 7.6, I have built a livepatch for the upstream fix for CVE-2018-9568 using kpatch rev. b29227547d6e.
A modified reproducer program from that commit can be used to trigger `sk_clone_lock()` ([test-clone.c.txt](https://github.com/dynup/kpatch/files/2691314/test-clone.c.txt))

    # kpatch load ./livepatch-CVE-2018-9568.ko 
    loading patch module: ./livepatch-CVE-2018-9568.ko
    waiting (up to 5 seconds) for patch transition to complete...
    transition complete (1 seconds)

    # mkdir /sys/fs/cgroup/memory/my_test/
    # echo 655360 > /sys/fs/cgroup/memory/my_test/memory.kmem.tcp.limit_in_bytes
    # sh
    # echo $$ > /sys/fs/cgroup/memory/my_test/tasks
    # /home/user/temp/test-clone

    (kernel crashed at this point)

`WARNING: CPU: 1 PID: 4004 at mm/page_counter.c:29 page_counter_cancel+0x42/0x50`, followed by a `BUG: unable to handle kernel paging request`, probably due to memory corruption.

mm/page_counter.c:29:
```
int page_counter_cancel(struct page_counter *counter, unsigned long nr_pages)
{
	long new;

	new = atomic_long_sub_return(nr_pages, &counter->count);

	/* More uncharges than charges? */
	WARN_ON_ONCE(new < 0);   // here

	return new > 0;
}
```
The full dmesg is attached: [vmcore-dmesg.txt](https://github.com/dynup/kpatch/files/2691312/vmcore-dmesg.txt)

---

A workaround for that particular patch is simple: just patch `sk_update_clone()` too, so that it checked the value of the static key explicitly:

```
--- a/net/core/sock.c
+++ b/net/core/sock.c
@@ -1470,7 +1470,7 @@ EXPORT_SYMBOL(sk_release_kernel);
 
 static void sk_update_clone(const struct sock *sk, struct sock *newsk)
 {
-	if (mem_cgroup_sockets_enabled && sk->sk_cgrp)
+	if (static_key_enabled(&memcg_socket_limit_enabled) && sk->sk_cgrp)
 		sock_update_memcg(newsk);
 }
```

However, I think, such things in the source patches are almost impossible to find during the code review. It would be better if kpatch-build could at least detect such conditions and fail the build of the patch or build the patch so that the sure static keys were handled automatically.

---

Files:
[test-clone.c.txt](https://github.com/dynup/kpatch/files/2691314/test-clone.c.txt)
[vmcore-dmesg.txt](https://github.com/dynup/kpatch/files/2691312/vmcore-dmesg.txt)

**Comment by ufo19890607** [comment]
I've also found the similar issue when I added hot patch in schedule funcs , like, 'check_preempt_wakeup', and cfs_bandwidth_used will never return true because it invokes static_key_false(&__cfs_bandwidth_used).

**Comment by euspectre** [comment]
kpatch-build/create-diff-object.c, kpatch_process_special_sections():

```c
	/*
	 * The following special sections aren't supported, so make sure we
	 * don't ever try to include them.  Otherwise the kernel will see the
	 * jump table during module loading and get confused.  Generally it
	 * should be safe to exclude them, it just means that you can't modify
	 * jump labels and enable tracepoints in a patched function.
	 */
	list_for_each_entry(sec, &kelf->sections, list) {
		if (strcmp(sec->name, "__jump_table") &&
		    strcmp(sec->name, "__tracepoints") &&
		    strcmp(sec->name, "__tracepoints_ptrs") &&
		    strcmp(sec->name, "__tracepoints_strings"))
			continue;
```

The locations of the static key conditions are stored in `__jump_table` sections. These sections are skipped in kpatch_process_special_sections(), that is why the static key conditions remain in their default state in the patch module.

Now, if there was a way to detect if a patch touches the functions with such conditions, kpatch-build could at least report an error and fail the build.

**Comment by euspectre** [comment]
Something like this (not tested yet):
```patch
diff --git a/kpatch-build/create-diff-object.c b/kpatch-build/create-diff-object.c
index e167890..fe4f11a 100644
--- a/kpatch-build/create-diff-object.c
+++ b/kpatch-build/create-diff-object.c
@@ -2375,6 +2375,21 @@ static void kpatch_create_kpatch_arch_section(struct kpatch_elf *kelf, char *obj
 	karch_sec->sh.sh_size = karch_sec->data->d_size;
 }
 
+static void check_jump_labels(struct section *sec)
+{
+	struct rela *rela;
+
+	if (!sec->rela)
+		return;
+
+	list_for_each_entry(rela, &sec->rela->relas, list) {
+		if (!rela->sym->name || rela->sym->status == SAME)
+			continue;
+		ERROR("the patch changes function \"%s\" which contains jump labels\n",
+		      rela->sym->name);
+	}
+}
+
 static void kpatch_process_special_sections(struct kpatch_elf *kelf)
 {
 	struct special_section *special;
@@ -2443,6 +2458,9 @@ static void kpatch_process_special_sections(struct kpatch_elf *kelf)
 		    strcmp(sec->name, "__tracepoints_strings"))
 			continue;
 
+		if (!strcmp(sec->name, "__jump_table"))
+			check_jump_labels(sec);
+
 		sec->status = SAME;
 		sec->include = 0;
 	
```

**Comment by euspectre** [comment]
Perhaps, "ERROR" is overkill here. This way, kpatch-build would complain about the functions with tracepoints as well.

**Comment by joe-lawrence** [comment]
We could maybe just log a notice for each instance of a skipped, unsupported section:

```patch
diff --git a/kpatch-build/create-diff-object.c b/kpatch-build/create-diff-object.c
index b4e3308..bae06f0 100644
--- a/kpatch-build/create-diff-object.c
+++ b/kpatch-build/create-diff-object.c
@@ -2449,6 +2449,13 @@ static void kpatch_process_special_sections(struct kpatch_elf *kelf)
 		if (sec->rela) {
 			sec->rela->status = SAME;
 			sec->rela->include = 0;
+
+			list_for_each_entry(rela, &sec->rela->relas, list) {
+				if (rela->sym->status == SAME)
+					continue;
+				log_normal("NOTICE: function \"%s\" changed with unsupported %s section\n",
+					   rela->sym->name, sec->name);
+			}
 		}
 	}
```

**Comment by ufo19890607** [comment]
Hi joe
Can you simply explain why kpatch does not support jump table, tracepoint?

**Comment by ufo19890607** [comment]
And we should also update 'Patch Author Guide'.

**Comment by euspectre** [comment]
> We could maybe just log a notice for each instance of a skipped, unsupported section:

Perhaps. I'd also output the offset of the location in the patched/new function to make it clearer where the jump label site is. The patch author may need to rewrite the static key condition to make things work like I mentioned earlier.

My recent experiments have shown that there are quite many jump labels in the code touched by security fixes I have built livepatches for. Most of these likely belong to tracepoints, which might be OK to leave disabled. But it is needed to inspect the jump labels anyway, to play safer.

**Comment by jpoimboe** [comment]
We should definitely add support for jump labels.  I don't think it should be too hard.  I'll see what I can do.

**Comment by jpoimboe** [comment]
@euspectre Can you try jpoimboe@74a7f922ad5500ce89cf8cfab82099c448eb3e7c (completely untested) ?

**Comment by euspectre** [comment]
> Can you try [jpoimboe/kpatch@ed04e50](https://github.com/jpoimboe/kpatch/commit/ed04e508eee478a7abfbaa0a54f6e749f6c8b2be) (completely untested) ?

Thanks, I'll try it, hopefully, in a day or two.

**Comment by euspectre** [comment]
@jpoimboe: The patch seems to be working OK. Thanks!
I checked it again in CentOS 7 with kernel 3.10.0-957.1.3.el7 and the same example as above.

When I enabled memory.kmem.tcp.limit_in_bytes, the kernel updated the binary code of the patched function appropriately (nop => cmp). And it switched it back to nop after the limit has been disabled.

No crashes so far, of course.
